### PR TITLE
CLN: groupby.ops follow-up cleanup

### PR DESCRIPTION
**Pull request by jbrockmendel**
get the namespaces/privateness "right"
avoid runtime imports
fix "type: ignore"s

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -261,7 +274,15 @@ def get_out_dtype(self, dtype: np.dtype) -> np.dtype:
                 out_dtype = "object"
         return np.dtype(out_dtype)
 
-    def get_result_dtype(self, dtype: DtypeObj) -> DtypeObj:
+    @overload
+    def _get_result_dtype(self, dtype: np.dtype) -> np.dtype:
+        ...
+
+    @overload
+    def _get_result_dtype(self, dtype: ExtensionDtype) -> ExtensionDtype:
```
Comment (mzeitlin11) [inline pandas/core/groupby/ops.py:282]: Would a TypeVar work here? Might end up being useful elsewhere too

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:282]: i dont think that would be accurate here

Comment (mzeitlin11) [reply pandas/core/groupby/ops.py:282]: Might be misunderstanding overloads, but is this expressing `ExtensionDtype -> ExtensionDtype` and `np.dtype -> np.dtype`? 

So was thinking a TypeVar like `TypeVar('DtypeObjT', np.dtype, ExtensionDtype)`

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:282]: do TypeVars not preserve subclasses?

Comment (mzeitlin11) [reply pandas/core/groupby/ops.py:282]: Running `mypy` on something like

```
from typing import TypeVar

AnyStr = TypeVar('AnyStr', str, bytes)


class Upper(str):
    def __new__(cls, val):
        return str.__new__(cls, val.upper())


class Lower(str):
    def __new__(cls, val):
        return str.__new__(cls, val.lower())


def swap_lower_upper(s: AnyStr) -> AnyStr:
    if isinstance(s, bytes):
        return s
    elif isinstance(s, Lower):
        return Upper(s)
    else:
        return Lower(s)

swap_lower_upper(Lower("hello"))
swap_lower_upper(Upper("hello"))
```
gives no error, so I don't think subclasses are preserved.

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:282]: cool.  will make a separate PR to implement DtypeObjT

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py:282]: implenting this with DtypeObjT and getting rid of the overloads im seeing

```
error: Incompatible return value type (got "dtype[signedinteger[_64Bit]]", expected "ExtensionDtype")  [return-value]
```

Comment (mzeitlin11) [reply pandas/core/groupby/ops.py:282]: Will have time to look in more detail later, initial guess would be that this return is the problem https://github.com/pandas-dev/pandas/blob/59b2db14f9c47a1b34da9efc12d708d43bcc893f/pandas/core/groupby/ops.py#L288-L290

IIRC mypy can't narrow based on type equality, so it's thinking that an `ExtensionDtype` can hit that return. But then not sure why it wouldn't complain with the overload, but those are confusing :)

**Review by jreback — APPROVED** [review]
looks fine

cc @simonjayhawkins if comments

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel lgtm. TypeVars with a union of types do have different semantics than the `bound=` variant that we normally use, so can address @mzeitlin11 comments here or as follow-up.

**Pull request merged**